library/std/src/sys_common/thread_parker/wait_flag.rs

```diff
@@ -0,0 +1,96 @@
+//!    currently waiting on it.
+//! * `PARKED` means the token is not available and the thread is parked.
+//! * `NOTIFIED` means the token is available.
+//!
+//! `park` and `park_timeout` change the state from `EMPTY` to `PARKED` and from
+//! `NOTIFIED` to `EMPTY`. If the state was `NOTIFIED`, the thread was unparked and
+//! execution can continue without calling into the OS. If the state was `EMPTY`,
+//! the token is not available and the thread waits on the primitive (here called
+//! "wait flag").
+//!
+//! `unpark` changes the state to `NOTIFIED`. If the state was `PARKED`, the thread
+//! is or will be sleeping on the wait flag, so we raise it. Only the first thread
+//! to call `unpark` will raise the wait flag, so spurious wakeups are avoided
+//! (this is especially important for semaphores).
+
+use crate::pin::Pin;
+use crate::sync::atomic::AtomicI8;
+use crate::sync::atomic::Ordering::SeqCst;
+use crate::sys::wait_flag::WaitFlag;
+use crate::time::Duration;
+
+const EMPTY: i8 = 0;
+const PARKED: i8 = -1;
+const NOTIFIED: i8 = 1;
+
+pub struct Parker {
+    state: AtomicI8,
+    wait_flag: WaitFlag,
+}
+
+impl Parker {
+    /// Construct a parker for the current thread. The UNIX parker
+    /// implementation requires this to happen in-place.
+    pub unsafe fn new(parker: *mut Parker) {
+        parker.write(Parker { state: AtomicI8::new(EMPTY), wait_flag: WaitFlag::new() })
+    }
+
+    // This implementation doesn't require `unsafe` and `Pin`, but other implementations do.
+    pub unsafe fn park(self: Pin<&Self>) {
+        // The state values are chosen so that this subtraction changes
+        // `NOTIFIED` to `EMPTY` and `EMPTY` to `PARKED`.
+        let state = self.state.fetch_sub(1, SeqCst);
+        match state {
+            EMPTY => (),
+            NOTIFIED => return,
+            _ => panic!("inconsistent park state"),
+        }
+
+        self.wait_flag.wait();
+
+        // We need to do a load here to use `Acquire` ordering.
+        self.state.swap(EMPTY, SeqCst);
+    }
+
+    // This implementation doesn't require `unsafe` and `Pin`, but other implementations do.
+    pub unsafe fn park_timeout(self: Pin<&Self>, dur: Duration) {
+        let state = self.state.fetch_sub(1, SeqCst);
+        match state {
+            EMPTY => (),
+            NOTIFIED => return,
+            _ => panic!("inconsistent park state"),
+        }
+
+        let wakeup = self.wait_flag.wait_timeout(dur);
+        let state = self.state.swap(EMPTY, SeqCst);
+        if state == NOTIFIED && !wakeup {
+            // The token was made available after the wait timed out, but before
+            // we reset the state, so we need to reset the wait flag to avoid
+            // spurious wakeups. This wait has no timeout, but we know it will
+            // return quickly, as the unparking thread will definitely raise the
+            // flag if it has not already done so.
+            self.wait_flag.wait();
+        }
```

Comment:
After thinking for a while, this needs to be disabled at least for the SOLID target because of the possibility of unbounded priority inversion in priority-based scheduling.

Consider two threads `T_hi` and `T_lo` with different priorities, which issue `park_timeout` and `unpark(T_hi)` respectively.  If a third middle-priority thread `T_mid` preempts `T_lo` in the middle of `unpark`, `T_hi` may be blocked for an extended period of time, exceeding the specified timeout.

Reply:
Something like this might work for all targets:

```rust
if wakeup {
    // NOTIFIED → EMPTY (consume the token)
    self.state.swap(EMPTY, Acquire);
} else {
    // PARKED → EMPTY (the token was unavailable)
    // NOTIFIED → NOTIFIED (the token was made available after the timeout, but
    // we aren't going to consume it this time)
    self.state.compare_exchange(PARKED, EMPTY, Relaxed, Relaxed);
}
```

Reply:
That would not solve the problem of spurious wakeups, however, as if the state is `NOTIFIED` we never touch the event flag. I could change the `Parker` to be just a simple wrapper over the event flag (it has exactly the right semantics), as we don't actually guarantee any atomic synchronization, even though all other platforms do it. That would harm the case where the token is already available, however.

Reply:
Ah, I think I understand better now - [my last code](https://github.com/rust-lang/rust/pull/97140#discussion_r877788051) would leave `(wait_flag, state)` at `(true, NOTIFIED)`, which requires two `park` calls to consume completely, aggravating the problem.

Back to the original code (https://github.com/rust-lang/rust/pull/97140/commits/3b6ae15058dbb68710f92697265580c7e957629f): Doing `wait` here is safe on a platform that does not use strict priority-based scheduling but can still cause context switches in rare cases, which is likely to be more costly than the spurious wakeup that would occur otherwise. How about replacing the `wait()` with `try_wait()`¹ or `wait_timeout(Duration::ZERO)`? This way we can consume the excess token only if it's beneficial.

¹ For μITRON eventflags, this would be [`pol_flg`](http://www.ertl.jp/ITRON/SPEC/FILE/mitron-400e.pdf#page=160).

Reply:
> How about replacing the `wait()` with `try_wait()`¹ or `wait_timeout(Duration::ZERO)`? This way we can consume the excess token only if it's beneficial.

Implemented in <https://github.com/solid-rs/rust/commit/fbf68854bf8842913eacade50879b294977c85a8>. Feel free to incorporate it in the PR.

Reply:
I solved this the same way the futex-based implementation handles spurious wakeups: `park` spins in a loop until the token is actually set and `park_timeout` returns immediately.